File: web/app/types/costs.ts
// Coût LLM (#coût) — miroir de felix.cost.ModelCost / CostSummary, et de
// ProjectCostOut (GET /api/costs). `cost_usd` null = prix inconnu (jamais un
// faux 0, cf. felix.cost.DEFAULT_PRICING).

export interface ModelCostPayload {
  model: string
  request_tokens: number
  response_tokens: number
  total_tokens: number
  cost_usd: number | null
}

// Coût d'UNE opération (un tour de chat, un import de document) — porté par
// l'event SSE `usage` et par le payload `{ cost: ... }` des messages persistés.
export interface CostSummaryPayload {
  request_tokens: number
  response_tokens: number
  total_tokens: number
  cost_usd: number | null
  by_model: ModelCostPayload[]
}

// Total persistant du projet (GET /api/costs?project=), toutes opérations
// confondues depuis l'origine — indépendant de l'historique de conversation.
export interface ProjectCostPayload {
  request_tokens: number
  response_tokens: number
  total_tokens: number
  cost_usd: number | null
  count_by_kind: Record<string, number>
}

File: src/felix/api/routes/costs.py
"""Route du coût persistant par projet (#coût) — total agrégé (tokens + USD)
et compte d'opérations par nature (`chat`/`ingest`), depuis les nœuds
`:CostEntry` posés à chaque tour de chat / import de document (cf.
`felix.core.costs`). Consommée par la topbar du front (total du projet,
rafraîchi après chaque tour/import) — INDÉPENDANT de l'historique des
messages, survit à « nouvelle conversation »/archivage.
"""

from __future__ import annotations

from fastapi import APIRouter
from pydantic import BaseModel

from felix.api.deps import Neo4jDriver  # noqa: TC001 — FastAPI résout au runtime (DI)
from felix.core.costs import project_cost_totals
from felix.core.projects import DEFAULT_PROJECT

router = APIRouter(prefix="/api/costs", tags=["costs"])


class ProjectCostOut(BaseModel):
    request_tokens: int
    response_tokens: int
    total_tokens: int
    cost_usd: float | None
    count_by_kind: dict[str, int]


@router.get("")
async def get_project_costs(
    driver: Neo4jDriver, project: str = DEFAULT_PROJECT
) -> ProjectCostOut:
    totals = await project_cost_totals(driver, project=project)
    return ProjectCostOut(**totals)
